includes/wc-update-functions.php

```diff
@@ -1773,3 +1773,123 @@ function wc_update_340_last_active() {
 function wc_update_340_db_version() {
 	WC_Install::update_db_version( '3.4.0' );
 }
+
+/**
+ * Copy order customer_id from post meta to post_author and set post_author to 0 for refunds.
+ *
+ * Two different strategies are used to copy data depending if the update is being executed from
+ * the command line or not. If `wp wc update` is used to update the database, this function
+ * copies data in a single go that is faster but uses more resources. If the databse update was
+ * triggered from the wp-admin, this function copies data in batches which is slower but uses
+ * few resources and thus is less likely to fail on smaller servers.
+ *
+ * @param WC_Background_Updater|false $updater Background updater instance or false if function is called from `wp wc update` WP-CLI command.
+ */
+function wc_update_350_order_customer_id( $updater = false ) {
+	global $wpdb;
+
+	$post_types              = (array) apply_filters( 'woocommerce_update_350_order_customer_id_post_types', array( 'shop_order' ) );
+	$post_types_placeholders = implode( ', ', array_fill( 0, count( $post_types ), '%s' ) );
+
+	if ( defined( 'WP_CLI' ) && WP_CLI ) {
+		// If running the update from the command-line, copy data in a single go which is faster but uses more resources.
+		$wpdb->query(
+			'CREATE TEMPORARY TABLE customers_map (post_id BIGINT(20), customer_id BIGINT(20), PRIMARY KEY(post_id))'
+		);
+
+		$wpdb->query(
+			"INSERT IGNORE INTO customers_map (SELECT post_id, meta_value FROM wp_postmeta WHERE meta_key = '_customer_user')"
+		);
+
+		$wpdb->query( 'SET sql_safe_updates=1' );
+
+		$wpdb->query(
+			$wpdb->prepare(
+				"UPDATE wp_posts JOIN customers_map ON wp_posts.ID = customers_map.post_id SET wp_posts.post_author = customers_map.customer_id WHERE post_type IN ({$post_types_placeholders})",  // phpcs:ignore WordPress.WP.PreparedSQL.NotPrepared
+				$post_types
+			)
+		);
+
+		$wpdb->update( $wpdb->posts, array( 'post_author' => 0 ), array( 'post_type' => 'shop_order_refund' ) );
+	} else {
+		// If running the update from the wp-admin, copy data in batches being careful not to use more memory than allowed.
+		$admin_orders_sql = '';
+		$admin_orders     = get_transient( 'wc_update_350_admin_orders' );
+
+		if ( false === $admin_orders ) {
+			// Get the list of orders that we don't want to change as they belong to user ID 1.
+			$admin_orders = $wpdb->get_col(
+				$wpdb->prepare(
+					"SELECT ID FROM wp_posts p
+					INNER JOIN wp_postmeta pm ON p.ID = pm.post_id
+					WHERE post_type IN ({$post_types_placeholders}) AND meta_key = '_customer_user' AND meta_value = 1", // phpcs:ignore WordPress.WP.PreparedSQL.NotPrepared
+					$post_types
+				)
+			);
+
+			// Cache the list of orders placed by the user ID 1 as to large stores this query can be slow.
+			set_transient( 'wc_update_350_admin_orders', $admin_orders, DAY_IN_SECONDS );
```

Comment:
Is this more efficient than a routine to first set all post author fields to 0 before doing this?

Reply:
I remember I considered updating all post author fields to 0 but to be honest I don't remember why I chose another approach. One thing to consider is that in the when running the update via the wp-admin I'm not trying to be as efficient as possible but instead to be mindful of resources consumption. Updating all post_author fields of all orders at once is likely to consume more resources than getting the list of orders that belong to the admin user.

Reply:
Even if thats a pure sql thing? The method you're using involves PHP loops, and if we're missing something (size of admin orders) it could cause resource issues itself.

Reply:
After running some tests, I now remember why I decided to get a list of orders that belong to the admin instead of updating all post_author fields to 0. It seems to me that updating all post_author fields to 0 will lock wp_posts to inserts and updates while the query is running which can be problematic for large and busy stores as the query will take some time to complete while the site is live. Queries could timeout while waiting for a lock. The same wouldn't happen with the select query to get all the orders placed by the admin.

It is important to mention that I don't have a deep understanding of how InnoDB locks works, so there is a chance I'm missing something.